docs/security.rst

```diff
@@ -210,7 +220,99 @@ permission check will fail.
    ...    [subscription.isApplicable(event.object) for subscription in subscriptions]
    [False]
 
+Automatic Deactivation On Failure
+=================================
+
+For any type of subscription (static, dynamic, persistent, ...) that
+implements
+:class:`nti.webhooks.interfaces.ILimitedApplicabilityPreconditionFailureWebhookSubscription`,
+attempting to make deliveries to it while it is misconfigured (e.g.,
+there is no such permission defined or no principal can be found) will
+eventually result in it becoming automatically disabled.
+
+Our subscription is such an object:
+
+.. doctest::
+
+   >>> from nti.webhooks.interfaces import ILimitedApplicabilityPreconditionFailureWebhookSubscription
+   >>> from zope.interface import verify
+   >>> subscription = subscriptions[0]
+   >>> verify.verifyObject(ILimitedApplicabilityPreconditionFailureWebhookSubscription, subscription)
+   True
+   >>> subscription.applicable_precondition_failure_limit
+   50
+   >>> subscription.active
+   True
+   >>> print(subscription.status_message)
+   Active
+
+This doesn't apply when the permission check is simply denied; that's
+normal and expected.
+
+.. doctest::
+
+   >>> from delivery_helper import deliver_some
+   >>> from nti.webhooks.testing import wait_for_deliveries
+   >>> with interaction('some.one.else'):
+   ...      deliver_some(100)
+   >>> len(subscription)
+   0
+   >>> subscription.active
+   True
+
+But if we remove the principal our subscription is using (being
+careful not to fire any events that might automatically remove or
+deactivate the subscription) we can see that it becomes inactive at
+the correct time.
+
```

Comment:
Can you imagine facilities to migrate subscriptions from one principal to another? I'm imagining what this looks like as users come and go from organizations?

To some extent many webhooks are really a function of the site administration, where there may be a set of principals acting as admins for a system. I wonder if these things being principal specific is going to become an issue for our use case.

Reply:
I can imagine facilities to change owners, yeah, pretty easy. There's a TODO about that somewhere, or at least a very specific warning not to try to do that right now. It wouldn't be hard.

>  I wonder if these things being principal specific is going to become an issue for our use case.

As far as Zapier goes, the very first note I wrote down is that "Zaps are oriented to **individual** end users. But teams can share connected apps." And later, more about the shared, connected apps: they are still tied to an individual user, and that user's credentials. If you have to change that, there's a process you have to go through on their end.